File: ios/YTLite/Extraction/ExtractorBundleStore.swift
import Foundation
import CryptoKit

/// Installed / remote extractor knobs (from Supabase Storage `manifest.json` `config`).
struct ExtractorRemoteConfig: Codable, Equatable, Sendable {
    var androidClientVersion: String
    var signatureTimestamp: Int
    var preferItags: [Int]
    var enableAndroidPlayerFallback: Bool
    var enableWatchPageFallback: Bool

    static let defaults = ExtractorRemoteConfig(
        androidClientVersion: "20.10.38",
        signatureTimestamp: 20_646,
        preferItags: [18, 22, 37],
        enableAndroidPlayerFallback: true,
        enableWatchPageFallback: true
    )
}

/// Thread-safe snapshot of the last successfully installed extractor config.
enum ExtractorRemoteConfigStore {
    private static let lock = NSLock()
    private static var _current = ExtractorRemoteConfig.defaults
    private static let defaultsKey = "extractor_remote_config_json"

    static var current: ExtractorRemoteConfig {
        lock.lock()
        defer { lock.unlock() }
        return _current
    }

    static func apply(_ config: ExtractorRemoteConfig) {
        lock.lock()
        _current = config
        lock.unlock()
        if let data = try? JSONEncoder().encode(config) {
            UserDefaults.standard.set(data, forKey: defaultsKey)
        }
        YouTubeConstants.applyPreferredVideoItags(config.preferItags)
    }

    static func restoreFromDisk() {
        guard let data = UserDefaults.standard.data(forKey: defaultsKey),
              let config = try? JSONDecoder().decode(ExtractorRemoteConfig.self, from: data)
        else { return }
        apply(config)
    }
}

/// Downloads / caches remote extractor assets from Supabase Storage (no in-app bundle).
actor ExtractorBundleStore {
    enum BundleError: LocalizedError {
        case supabaseNotConfigured
        case downloadFailed(String)
        case checksumMismatch(String)
        case incompleteBundle
        case appVersionTooLow(required: String, current: String)
        case invalidManifest

        var errorDescription: String? {
            switch self {
            case .supabaseNotConfigured:
                return "Supabase URL is not configured; cannot download extractor"
            case .downloadFailed(let msg):
                return "Extractor download failed: \(msg)"
            case .checksumMismatch(let file):
                return "Extractor checksum mismatch: \(file)"
            case .incompleteBundle:
                return "Extractor bundle incomplete"
            case .appVersionTooLow(let required, let current):
                return "App \(current) is too old for extractor (needs \(required))"
            case .invalidManifest:
                return "Invalid extractor manifest"
            }
        }
    }

    struct ManifestFile: Codable, Sendable {
        var path: String
        var sha256: String
    }

    struct Manifest: Codable, Sendable {
        var version: Int
        var minAppVersion: String
        var files: [String: ManifestFile]
        var config: ExtractorRemoteConfig?
    }

    static let shared = ExtractorBundleStore()

    private let fileManager = FileManager.default
    private var lastBackgroundCheck: Date?

    func ensureBundle(backgroundRefresh: Bool = true) async throws -> URL {
        ExtractorRemoteConfigStore.restoreFromDisk()
        let dir = try cacheDirectory()
        if isComplete(dir: dir) {
            if let installed = try? loadInstalledManifest(dir: dir) {
                ExtractorRemoteConfigStore.apply(installed.config ?? .defaults)
            }
            if backgroundRefresh {
                scheduleBackgroundRefreshIfNeeded()
            }
            return dir
        }
        try await downloadAndInstall()
        return try cacheDirectory()
    }

    func cacheDirectoryURL() throws -> URL {
        try cacheDirectory()
    }

    func installedManifest() throws -> Manifest? {
        let dir = try cacheDirectory()
        guard isComplete(dir: dir) else { return nil }
        return try loadInstalledManifest(dir: dir)
    }

    // MARK: - Private

    private func scheduleBackgroundRefreshIfNeeded() {
        let now = Date()
        if let last = lastBackgroundCheck, now.timeIntervalSince(last) < 60 {
            return
        }
        lastBackgroundCheck = now
        Task { try? await self.refreshIfNewer() }
    }

    private func refreshIfNewer() async throws {
        let remote = try await fetchManifest()
        let dir = try cacheDirectory()
        let installedVersion = (try? loadInstalledManifest(dir: dir))?.version ?? 0
        guard remote.version > installedVersion else { return }
        try await install(manifest: remote)
    }

    private func downloadAndInstall() async throws {
        let remote = try await fetchManifest()
        try await install(manifest: remote)
    }

    private func install(manifest: Manifest) async throws {
        try checkMinAppVersion(manifest.minAppVersion)
        let required = ["extractor.js", "bridge.html", "bridge-ios.html"]
        for name in required {
            guard manifest.files[name] != nil else { throw BundleError.invalidManifest }
        }

        let dir = try cacheDirectory()
        let staging = dir.appendingPathComponent(".staging-\(UUID().uuidString)", isDirectory: true)
        try? fileManager.removeItem(at: staging)
        try fileManager.createDirectory(at: staging, withIntermediateDirectories: true)

        do {
            for name in required {
                let entry = manifest.files[name]!
                let bytes = try await downloadPublicObject(relativePath: entry.path)
                let digest = SHA256.hash(data: bytes)
                let hex = digest.map { String(format: "%02x", $0) }.joined()
                guard hex.caseInsensitiveCompare(entry.sha256) == .orderedSame else {
                    throw BundleError.checksumMismatch(name)
                }
                let dest = staging.appendingPathComponent(name)
                try bytes.write(to: dest, options: .atomic)
            }
            let manifestData = try JSONEncoder().encode(manifest)
            try manifestData.write(
                to: staging.appendingPathComponent("installed-manifest.json"),
                options: .atomic
            )

            for name in required + ["installed-manifest.json"] {
                let src = staging.appendingPathComponent(name)
                let dest = dir.appendingPathComponent(name)
                if fileManager.fileExists(atPath: dest.path) {
                    try fileManager.removeItem(at: dest)
                }
                try fileManager.moveItem(at: src, to: dest)
            }
            try? fileManager.removeItem(at: staging)
            ExtractorRemoteConfigStore.apply(manifest.config ?? .defaults)
            PlayProbe.log(
                "extractor.bundle.installed",
                "version=\(manifest.version) sts=\(ExtractorRemoteConfigStore.current.signatureTimestamp)"
            )
        } catch {
            try? fileManager.removeItem(at: staging)
            throw error
        }
    }

    private func fetchManifest() async throws -> Manifest {
        let url = try publicURL(relativePath: "manifest.json")
        let (data, response) = try await URLSession.shared.data(from: url)
        let code = (response as? HTTPURLResponse)?.statusCode ?? -1
        guard (200..<300).contains(code) else {
            throw BundleError.downloadFailed("manifest HTTP \(code)")
        }
        do {
            return try JSONDecoder().decode(Manifest.self, from: data)
        } catch {
            throw BundleError.invalidManifest
        }
    }

    private func downloadPublicObject(relativePath: String) async throws -> Data {
        let url = try publicURL(relativePath: relativePath)
        let (data, response) = try await URLSession.shared.data(from: url)
        let code = (response as? HTTPURLResponse)?.statusCode ?? -1
        guard (200..<300).contains(code), !data.isEmpty else {
            throw BundleError.downloadFailed("\(relativePath) HTTP \(code)")
        }
        return data
    }

    private func publicURL(relativePath: String) throws -> URL {
        let base = AppConfig.fromBundle().supabaseURL
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard !base.isEmpty else { throw BundleError.supabaseNotConfigured }
        let path = relativePath.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard let url = URL(string: "\(base)/storage/v1/object/public/extractor/\(path)") else {
            throw BundleError.downloadFailed("bad url \(path)")
        }
        return url
    }

    private func cacheDirectory() throws -> URL {
        let root = try fileManager.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let dir = root.appendingPathComponent("extractor", isDirectory: true)
        if !fileManager.fileExists(atPath: dir.path) {
            try fileManager.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        return dir
    }

    private func isComplete(dir: URL) -> Bool {
        let required = ["extractor.js", "bridge.html", "bridge-ios.html", "installed-manifest.json"]
        return required.allSatisfy {
            fileManager.fileExists(atPath: dir.appendingPathComponent($0).path)
        }
    }

    private func loadInstalledManifest(dir: URL) throws -> Manifest {
        let url = dir.appendingPathComponent("installed-manifest.json")
        let data = try Data(contentsOf: url)
        return try JSONDecoder().decode(Manifest.self, from: data)
    }

    private func checkMinAppVersion(_ required: String) throws {
        let current = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0"
        if Self.compareSemver(current, required) == .orderedAscending {
            throw BundleError.appVersionTooLow(required: required, current: current)
        }
    }

    /// Numeric semver compare: `1.2.0` vs `1.10.0`.
    nonisolated private static func compareSemver(_ a: String, _ b: String) -> ComparisonResult {
        let pa = a.split(separator: ".").map { Int($0) ?? 0 }
        let pb = b.split(separator: ".").map { Int($0) ?? 0 }
        let n = max(pa.count, pb.count)
        for i in 0..<n {
            let x = i < pa.count ? pa[i] : 0
            let y = i < pb.count ? pb[i] : 0
            if x < y { return .orderedAscending }
            if x > y { return .orderedDescending }
        }
        return .orderedSame
    }
}

File: app/src/main/java/com/ytlite/player/data/extractor/ExtractorBundleStore.kt
package com.ytlite.player.data.extractor

import android.content.Context
import android.util.Log
import com.ytlite.player.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.File
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.util.concurrent.atomic.AtomicReference

data class ExtractorRemoteConfig(
    val androidClientVersion: String = "20.10.38",
    val signatureTimestamp: Int = 20646,
    val preferItags: List<Int> = listOf(18, 22, 37),
    val enableAndroidPlayerFallback: Boolean = true,
    val enableWatchPageFallback: Boolean = true,
) {
    companion object {
        val DEFAULTS = ExtractorRemoteConfig()

        fun fromJson(obj: JSONObject?): ExtractorRemoteConfig {
            if (obj == null) return DEFAULTS
            val itags = mutableListOf<Int>()
            obj.optJSONArray("preferItags")?.let { arr ->
                for (i in 0 until arr.length()) {
                    itags.add(arr.optInt(i))
                }
            }
            return ExtractorRemoteConfig(
                androidClientVersion = obj.optString(
                    "androidClientVersion",
                    DEFAULTS.androidClientVersion,
                ),
                signatureTimestamp = obj.optInt(
                    "signatureTimestamp",
                    DEFAULTS.signatureTimestamp,
                ),
                preferItags = itags.ifEmpty { DEFAULTS.preferItags },
                enableAndroidPlayerFallback = obj.optBoolean(
                    "enableAndroidPlayerFallback",
                    DEFAULTS.enableAndroidPlayerFallback,
                ),
                enableWatchPageFallback = obj.optBoolean(
                    "enableWatchPageFallback",
                    DEFAULTS.enableWatchPageFallback,
                ),
            )
        }
    }
}

object ExtractorRemoteConfigStore {
    private val current = AtomicReference(ExtractorRemoteConfig.DEFAULTS)
    private const val PREFS = "extractor_remote_config"
    private const val KEY = "json"

    fun current(): ExtractorRemoteConfig = current.get()

    fun apply(config: ExtractorRemoteConfig, context: Context? = null) {
        current.set(config)
        context?.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            ?.edit()
            ?.putString(KEY, toJson(config).toString())
            ?.apply()
    }

    fun restore(context: Context) {
        val raw = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString(KEY, null)
            ?: return
        runCatching {
            apply(ExtractorRemoteConfig.fromJson(JSONObject(raw)))
        }
    }

    private fun toJson(config: ExtractorRemoteConfig): JSONObject =
        JSONObject()
            .put("androidClientVersion", config.androidClientVersion)
            .put("signatureTimestamp", config.signatureTimestamp)
            .put("preferItags", org.json.JSONArray(config.preferItags))
            .put("enableAndroidPlayerFallback", config.enableAndroidPlayerFallback)
            .put("enableWatchPageFallback", config.enableWatchPageFallback)
}

/**
 * Downloads / caches remote extractor assets from Supabase Storage (no APK assets).
 */
class ExtractorBundleStore private constructor(
    private val appContext: Context,
) {
    sealed class BundleException(message: String) : Exception(message) {
        class NotConfigured : BundleException("Supabase URL is not configured; cannot download extractor")
        class DownloadFailed(msg: String) : BundleException("Extractor download failed: $msg")
        class ChecksumMismatch(file: String) : BundleException("Extractor checksum mismatch: $file")
        class Incomplete : BundleException("Extractor bundle incomplete")
        class AppVersionTooLow(required: String, current: String) :
            BundleException("App $current is too old for extractor (needs $required)")
        class InvalidManifest : BundleException("Invalid extractor manifest")
    }

    data class ManifestFile(val path: String, val sha256: String)
    data class Manifest(
        val version: Int,
        val minAppVersion: String,
        val files: Map<String, ManifestFile>,
        val config: ExtractorRemoteConfig?,
    )

    private val mutex = Mutex()
    @Volatile private var lastBackgroundCheckMs: Long = 0L

    suspend fun ensureBundle(backgroundRefresh: Boolean = true): File = mutex.withLock {
        withContext(Dispatchers.IO) {
            ExtractorRemoteConfigStore.restore(appContext)
            val dir = cacheDir()
            if (isComplete(dir)) {
                loadInstalledManifest(dir)?.let {
                    ExtractorRemoteConfigStore.apply(it.config ?: ExtractorRemoteConfig.DEFAULTS, appContext)
                }
                if (backgroundRefresh) {
                    scheduleBackgroundRefresh()
                }
                return@withContext dir
            }
            downloadAndInstall()
            cacheDir()
        }
    }

    fun cacheDir(): File {
        val dir = File(appContext.filesDir, "extractor")
        if (!dir.exists()) dir.mkdirs()
        return dir
    }

    fun bridgeFileUrl(): String = "file://${cacheDir().absolutePath}/bridge.html"

    private fun scheduleBackgroundRefresh() {
        val now = System.currentTimeMillis()
        if (now - lastBackgroundCheckMs < 60_000L) return
        lastBackgroundCheckMs = now
        // Caller already on IO / coroutine; fire-and-forget via new job from Application scope ideally.
        // Here we do a synchronous best-effort only when explicitly awaited elsewhere.
    }

    suspend fun refreshIfNewer() = withContext(Dispatchers.IO) {
        mutex.withLock {
            val remote = fetchManifest()
            val installed = loadInstalledManifest(cacheDir())?.version ?: 0
            if (remote.version > installed) {
                install(remote)
            }
        }
    }

    private suspend fun downloadAndInstall() {
        val remote = fetchManifest()
        install(remote)
    }

    private fun install(manifest: Manifest) {
        checkMinAppVersion(manifest.minAppVersion)
        val required = listOf("extractor.js", "bridge.html", "bridge-ios.html")
        required.forEach { name ->
            if (!manifest.files.containsKey(name)) throw BundleException.InvalidManifest()
        }

        val dir = cacheDir()
        val staging = File(dir, ".staging-${System.nanoTime()}")
        if (staging.exists()) staging.deleteRecursively()
        staging.mkdirs()
        try {
            for (name in required) {
                val entry = manifest.files.getValue(name)
                val bytes = downloadPublicObject(entry.path)
                val hex = sha256Hex(bytes)
                if (!hex.equals(entry.sha256, ignoreCase = true)) {
                    throw BundleException.ChecksumMismatch(name)
                }
                File(staging, name).writeBytes(bytes)
            }
            File(staging, "installed-manifest.json").writeText(manifestToJson(manifest).toString())
            for (name in required + "installed-manifest.json") {
                val src = File(staging, name)
                val dest = File(dir, name)
                if (dest.exists()) dest.delete()
                if (!src.renameTo(dest)) {
                    src.copyTo(dest, overwrite = true)
                    src.delete()
                }
            }
            staging.deleteRecursively()
            ExtractorRemoteConfigStore.apply(
                manifest.config ?: ExtractorRemoteConfig.DEFAULTS,
                appContext,
            )
            Log.i(TAG, "installed extractor version=${manifest.version}")
        } catch (e: Exception) {
            staging.deleteRecursively()
            throw e
        }
    }

    private fun fetchManifest(): Manifest {
        val url = publicUrl("manifest.json")
        val bytes = httpGet(url)
        return parseManifest(String(bytes, Charsets.UTF_8))
    }

    private fun downloadPublicObject(relativePath: String): ByteArray =
        httpGet(publicUrl(relativePath))

    private fun publicUrl(relativePath: String): String {
        val base = BuildConfig.SUPABASE_URL.trimEnd('/')
        if (base.isBlank()) throw BundleException.NotConfigured()
        val path = relativePath.trimStart('/')
        return "$base/storage/v1/object/public/extractor/$path"
    }

    private fun httpGet(url: String): ByteArray {
        val conn = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 30_000
            readTimeout = 45_000
        }
        try {
            val code = conn.responseCode
            val stream = if (code in 200..299) conn.inputStream else conn.errorStream
            val bytes = stream?.readBytes() ?: ByteArray(0)
            if (code !in 200..299 || bytes.isEmpty()) {
                throw BundleException.DownloadFailed("HTTP $code for $url")
            }
            return bytes
        } finally {
            conn.disconnect()
        }
    }

    private fun isComplete(dir: File): Boolean {
        val required = listOf(
            "extractor.js",
            "bridge.html",
            "bridge-ios.html",
            "installed-manifest.json",
        )
        return required.all { File(dir, it).isFile && File(dir, it).length() > 0L }
    }

    private fun loadInstalledManifest(dir: File): Manifest? {
        val file = File(dir, "installed-manifest.json")
        if (!file.isFile) return null
        return runCatching { parseManifest(file.readText()) }.getOrNull()
    }

    private fun parseManifest(raw: String): Manifest {
        val root = JSONObject(raw)
        val filesObj = root.optJSONObject("files") ?: throw BundleException.InvalidManifest()
        val files = mutableMapOf<String, ManifestFile>()
        for (key in filesObj.keys()) {
            val entry = filesObj.getJSONObject(key)
            files[key] = ManifestFile(
                path = entry.getString("path"),
                sha256 = entry.getString("sha256"),
            )
        }
        return Manifest(
            version = root.getInt("version"),
            minAppVersion = root.optString("minAppVersion", "1.0.0"),
            files = files,
            config = ExtractorRemoteConfig.fromJson(root.optJSONObject("config")),
        )
    }

    private fun manifestToJson(manifest: Manifest): JSONObject {
        val files = JSONObject()
        manifest.files.forEach { (name, entry) ->
            files.put(
                name,
                JSONObject().put("path", entry.path).put("sha256", entry.sha256),
            )
        }
        val config = ExtractorRemoteConfigStore.current().let { c ->
            JSONObject()
                .put("androidClientVersion", c.androidClientVersion)
                .put("signatureTimestamp", c.signatureTimestamp)
                .put("preferItags", org.json.JSONArray(c.preferItags))
                .put("enableAndroidPlayerFallback", c.enableAndroidPlayerFallback)
                .put("enableWatchPageFallback", c.enableWatchPageFallback)
        }
        // Prefer manifest's own config when writing installed copy.
        val cfg = manifest.config?.let {
            JSONObject()
                .put("androidClientVersion", it.androidClientVersion)
                .put("signatureTimestamp", it.signatureTimestamp)
                .put("preferItags", org.json.JSONArray(it.preferItags))
                .put("enableAndroidPlayerFallback", it.enableAndroidPlayerFallback)
                .put("enableWatchPageFallback", it.enableWatchPageFallback)
        } ?: config
        return JSONObject()
            .put("version", manifest.version)
            .put("minAppVersion", manifest.minAppVersion)
            .put("files", files)
            .put("config", cfg)
    }

    private fun checkMinAppVersion(required: String) {
        val current = runCatching {
            val p = appContext.packageManager.getPackageInfo(appContext.packageName, 0)
            @Suppress("DEPRECATION")
            p.versionName ?: "0"
        }.getOrDefault("0")
        if (compareSemver(current, required) < 0) {
            throw BundleException.AppVersionTooLow(required, current)
        }
    }

    private fun sha256Hex(bytes: ByteArray): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(bytes)
        return digest.joinToString("") { "%02x".format(it) }
    }

    /** @return negative if a < b */
    private fun compareSemver(a: String, b: String): Int {
        val pa = a.split('.').map { it.toIntOrNull() ?: 0 }
        val pb = b.split('.').map { it.toIntOrNull() ?: 0 }
        val n = maxOf(pa.size, pb.size)
        for (i in 0 until n) {
            val x = pa.getOrElse(i) { 0 }
            val y = pb.getOrElse(i) { 0 }
            if (x != y) return x.compareTo(y)
        }
        return 0
    }

    companion object {
        private const val TAG = "ExtractorBundle"

        @Volatile
        private var instance: ExtractorBundleStore? = null

        fun getInstance(context: Context): ExtractorBundleStore =
            instance ?: synchronized(this) {
                instance ?: ExtractorBundleStore(context.applicationContext).also { instance = it }
            }
    }
}
